src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/SignerInfo.cs

```diff
@@ -471,13 +472,26 @@ public void CheckSignature(X509Certificate2Collection extraStore, bool verifySig
 
         public void CheckHash()
         {
-            IncrementalHash hasher = PrepareDigest();
-            byte[] expectedSignature = hasher.GetHashAndReset();
+            if (CheckHash(compatMode: false) || CheckHash(compatMode: true))
+            {
+                return;
+            }
+
+            throw new CryptographicException(SR.Cryptography_BadSignature);
```

Comment:
Nit: I personally would find this easier to read as:
```C#
if (!CheckHash(compatMode: false) && !CheckHash(compatMode: true))
{
    throw new CryptographicException(SR.Cryptography_BadSignature);
}
```
but that's just personal preference.